src/test/java/build/buildfarm/instance/memory/MemoryInstanceTest.java

```diff
@@ -277,13 +277,58 @@ public Boolean answer(InvocationOnMock invocation) throws Throwable {
     verify(unfazedWatcher, times(1)).test(eq(doneOperation));
   }
 
-  private boolean putNovelOperation(Stage stage) throws InterruptedException {
-    return instance.putOperation(Operation.newBuilder()
-        .setName("does-not-exist")
+  @Test
+  public void requeueFailOnInvalid() throws InterruptedException, InvalidProtocolBufferException {
+    // These new operations are invalid as they're missing content.
+    Operation queuedOperation = createOperation("my-queued-operation", QUEUED);
+    outstandingOperations.put(queuedOperation.getName(), queuedOperation);
+
+    watchers.put(queuedOperation.getName(), new Predicate<Operation>() {
+      @Override
+      public boolean test(Operation operation) {
+        assertThat(operation.getError().getCode()).isEqualTo(Code.FAILED_PRECONDITION);
+        return false;
+      }
+    });
+
+    instance.requeueOperation(queuedOperation);
+
+    watchersThreadPool.shutdown();
+    watchersThreadPool.awaitTermination(Long.MAX_VALUE, TimeUnit.NANOSECONDS);
+  }
+
+  @Test
+  public void requeueFailOnNotQueued() throws InterruptedException, InvalidProtocolBufferException {
+    Operation queuedOperation = createOperation("my-queued-operation", QUEUED);
+    outstandingOperations.put(queuedOperation.getName(), queuedOperation);
+    Operation executingOperation = createOperation("my-executing-operation", EXECUTING);
+    outstandingOperations.put(executingOperation.getName(), executingOperation);
+
+    try {
+      instance.requeueOperation(executingOperation);
+      fail("Method should throw if operation is not QUEUED.");
+    } catch (IllegalStateException e) {
+      assertThat(e.getMessage()).isEqualTo("Operation stage is not QUEUED");
+    }
+
+    try {
+      instance.requeueOperation(queuedOperation);
+    } catch (IllegalStateException e) {
```

Comment:
this really isn't necessary, the method throwing is a substantial enough indicator that the test failed. Also, I'd much rather have the trace than a helpful message